File: packages/device_calendar_plus_android/android/src/main/kotlin/to/bullet/device_calendar_plus_android/PermissionService.kt
package to.bullet.device_calendar_plus_android

import android.Manifest
import android.app.Activity
import android.content.Context
import android.content.pm.PackageManager
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat

class PermissionService(private val context: Context) {

    private val activity: Activity?
        get() = context as? Activity
    
    companion object {
        const val CALENDAR_PERMISSION_REQUEST_CODE = 2024
        
        // Permission status values matching CalendarPermissionStatus enum
        const val STATUS_GRANTED = "granted"
        const val STATUS_WRITE_ONLY = "writeOnly"
        const val STATUS_DENIED = "denied"
        const val STATUS_NOT_DETERMINED = "notDetermined"

        private const val PREFS_PERMISSION_WAS_DENIED_BEFORE =
            "device_calendar_plus_permission_was_denied_before"
    }
    
    private var pendingCallback: ((Result<String>) -> Unit)? = null
    
    /**
     * Verifies the manifest declares the permissions the request needs.
     *
     * A write-only request only asks for (and therefore only requires)
     * [Manifest.permission.WRITE_CALENDAR], so an add-only app need not declare
     * [Manifest.permission.READ_CALENDAR]. A full request requires both.
     */
    private fun checkPermissionsDeclared(writeOnly: Boolean): PermissionException? {
        val readPermission = Manifest.permission.READ_CALENDAR
        val writePermission = Manifest.permission.WRITE_CALENDAR

        val packageInfo = context.packageManager.getPackageInfo(
            context.packageName,
            PackageManager.GET_PERMISSIONS
        )

        val declaredPermissions = packageInfo.requestedPermissions?.toList() ?: emptyList()

        val required = if (writeOnly) {
            listOf(writePermission)
        } else {
            listOf(readPermission, writePermission)
        }

        if (required.any { it !in declaredPermissions }) {
            val errorMessage = "Calendar permissions must be declared in AndroidManifest.xml.\n\n" +
                "Add the following to android/app/src/main/AndroidManifest.xml:\n" +
                required.joinToString("\n") { "<uses-permission android:name=\"$it\"/>" }

            return PermissionException(PlatformExceptionCodes.PERMISSIONS_NOT_DECLARED, errorMessage)
        }

        return null
    }

    /**
     * The access tier implied purely by which calendar permissions are
     * currently granted: [STATUS_GRANTED] (read + write), [STATUS_WRITE_ONLY]
     * (write alone — genuine add-only access, mirroring iOS 17+), or `null` when
     * write access isn't held. A `null` means there is no positive tier, so the
     * caller decides whether that reads as denied or not-yet-determined.
     */
    private fun grantedTier(): String? {
        val readGranted = ContextCompat.checkSelfPermission(
            context,
            Manifest.permission.READ_CALENDAR
        ) == PackageManager.PERMISSION_GRANTED

        val writeGranted = ContextCompat.checkSelfPermission(
            context,
            Manifest.permission.WRITE_CALENDAR
        ) == PackageManager.PERMISSION_GRANTED

        return when {
            readGranted && writeGranted -> STATUS_GRANTED
            writeGranted -> STATUS_WRITE_ONLY
            else -> null
        }
    }

    /**
     * Determines the current calendar permission status, distinguishing between
     * "never asked" ([STATUS_NOT_DETERMINED]) and "denied" ([STATUS_DENIED]).
     *
     * Android's [ContextCompat.checkSelfPermission] returns [PackageManager.PERMISSION_DENIED]
     * both when the user has never been asked and when they actively denied. To distinguish
     * these cases, we combine [ActivityCompat.shouldShowRequestPermissionRationale] with a
     * [android.content.SharedPreferences] flag that records whether a denial has occurred.
     *
     * The OS behavior (experimentally verified) follows this scenario table:
     *
     * | Previous state     | Action  | shouldShowRationale | SharedPrefs flag |
     * |--------------------|---------|---------------------|------------------|
     * | Not asked          | —       | false               | false            |
     * | Denied once        | Denied  | true                | true             |
     * | Denied once        | Dismiss | true                | true             |
     * | Permanently denied | Denied | false              | true             |
     *
     * Decision logic when write access isn't held (keyed off WRITE_CALENDAR
     * alone — the capability that defines whether any tier is reachable):
     * - `shouldShowRationale == false` AND SharedPrefs flag set --> [STATUS_DENIED] (permanently denied, must use app settings)
     * - everything else --> [STATUS_NOT_DETERMINED] (permission dialog can still be shown)
     *
     * Based on the approach from Baseflow's flutter-permission-handler:
     * https://github.com/Baseflow/flutter-permission-handler/blob/39fba431428e5d82d35f4999663461468fe3a728/permission_handler_android/android/src/main/java/com/baseflow/permissionhandler/PermissionUtils.java#L400-L536
     */
    private fun getCurrentPermissionStatus(): String {
        // Holding write access is a positive tier (full or write-only).
        grantedTier()?.let { return it }

        // Otherwise write isn't granted. Distinguish "never asked / can ask
        // again" (NOT_DETERMINED) from "permanently denied" (DENIED). WRITE is
        // the capability that defines whether any tier is reachable (a held
        // READ-only state isn't possible — grantedTier already returned for any
        // write-bearing tier), so the decision keys off WRITE alone.
        val writePermission = Manifest.permission.WRITE_CALENDAR

        // Without an Activity we can't check shouldShowRequestPermissionRationale,
        // so fall back to NOT_DETERMINED (safe default — caller can still request).
        val currentActivity = activity ?: return STATUS_NOT_DETERMINED

        val permanentlyDenied = wasPermissionDeniedBefore(writePermission) &&
            !ActivityCompat.shouldShowRequestPermissionRationale(currentActivity, writePermission)

        return if (permanentlyDenied) STATUS_DENIED else STATUS_NOT_DETERMINED
    }

    private fun wasPermissionDeniedBefore(permissionName: String): Boolean {
        val prefs = context.getSharedPreferences(permissionName, Context.MODE_PRIVATE)
        return prefs.getBoolean(PREFS_PERMISSION_WAS_DENIED_BEFORE, false)
    }

    private fun setPermissionDenied(permissionName: String) {
        val prefs = context.getSharedPreferences(permissionName, Context.MODE_PRIVATE)
        prefs.edit().putBoolean(PREFS_PERMISSION_WAS_DENIED_BEFORE, true).apply()
    }
    
    fun hasPermissions(): Result<String> {
        // A status check triggers no request, so only the minimal calendar
        // permission (WRITE) need be declared — an add-only app that omits
        // READ_CALENDAR can still check its status.
        val error = checkPermissionsDeclared(writeOnly = true)
        if (error != null) {
            return Result.failure(error)
        }

        return Result.success(getCurrentPermissionStatus())
    }
    
    /**
     * Requests calendar permissions from the user.
     *
     * @param writeOnly when `true`, requests only [Manifest.permission.WRITE_CALENDAR]
     *   (the add-only tier — mirrors iOS 17+ write-only). When `false`, requests
     *   both [Manifest.permission.READ_CALENDAR] and
     *   [Manifest.permission.WRITE_CALENDAR] for full access.
     */
    fun requestPermissions(writeOnly: Boolean, callback: (Result<String>) -> Unit) {
        val error = checkPermissionsDeclared(writeOnly)
        if (error != null) {
            callback(Result.failure(error))
            return
        }

        val currentStatus = getCurrentPermissionStatus()
        // Full access already satisfies any request. A write-only request is
        // also satisfied when write access is already held.
        if (currentStatus == STATUS_GRANTED ||
            (writeOnly && currentStatus == STATUS_WRITE_ONLY)) {
            callback(Result.success(currentStatus))
            return
        }

        val currentActivity = activity
        if (currentActivity == null) {
            callback(Result.failure(PermissionException(
                PlatformExceptionCodes.OPERATION_FAILED,
                "Cannot request permissions without an Activity. " +
                    "Use hasPermissions() to check status from a background context."
            )))
            return
        }

        // Store the callback to be completed when permission result is received
        pendingCallback = callback

        // Write-only asks for WRITE_CALENDAR alone; full access asks for both.
        // A full request while write-only is already held only re-prompts for
        // READ_CALENDAR, upgrading the tier.
        val readPermission = Manifest.permission.READ_CALENDAR
        val writePermission = Manifest.permission.WRITE_CALENDAR
        val requested = if (writeOnly) {
            arrayOf(writePermission)
        } else {
            arrayOf(readPermission, writePermission)
        }
        ActivityCompat.requestPermissions(
            currentActivity,
            requested,
            CALENDAR_PERMISSION_REQUEST_CODE
        )
    }
    
    fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ): Boolean {
        if (requestCode != CALENDAR_PERMISSION_REQUEST_CODE) {
            return false
        }
        
        val callback = pendingCallback ?: return false
        pendingCallback = null

        // A dismissed dialog delivers empty arrays — no grant, no denial. Report
        // the real current status (can-ask-again) instead of a hard denial, so a
        // later hasPermissions() doesn't disagree. Mirrors iOS, which re-reads
        // the real status on a dismissed prompt.
        if (grantResults.isEmpty()) {
            callback(Result.success(getCurrentPermissionStatus()))
            return true
        }

        // Record any denials so a later hasPermissions() can tell a permanent
        // denial from a can-ask-again one.
        permissions.forEachIndexed { index, permission ->
            if (grantResults.getOrNull(index) != PackageManager.PERMISSION_GRANTED) {
                setPermissionDenied(permission)
            }
        }

        // Report the tier actually held. WRITE_CALENDAR is the capability that
        // matters: hold it and the app has at least write-only access, so a
        // full request that granted write but denied read reads as writeOnly —
        // not denied — matching iOS, which reports the real tier on a non-full
        // grant. Lacking write, the request was denied. We use STATUS_DENIED
        // here (not getCurrentPermissionStatus's can-ask-again NOT_DETERMINED)
        // because a just-denied request should read as denied, again as on iOS.
        callback(Result.success(grantedTier() ?: STATUS_DENIED))
        return true
    }
}

data class PermissionException(
    val code: String,
    override val message: String
) : Exception(message)

File: packages/device_calendar_plus_ios/ios/device_calendar_plus_ios/Sources/device_calendar_plus_ios/PermissionService.swift
import EventKit
import Foundation

enum CalendarPermissionType {
  case write  // Need to write events (iOS 17+ writeOnly or fullAccess is fine)
  case full   // Need to read calendars/events (requires fullAccess)
}

/// What the OS reports about calendar access, normalised across iOS versions.
///
/// EventKit's own `EKAuthorizationStatus` means different things on either
/// side of iOS 17 (`.authorized` there is full access; `.writeOnly` does not
/// exist). Implementations flatten that away so `PermissionService` never has
/// to ask what OS it is on.
enum OSAuthorization {
  case notDetermined
  case denied
  case restricted
  case writeOnly
  case fullAccess
}

/// The OS-level calendar authorization seam.
///
/// Injected into `PermissionService` so tests can drive the whole grant path
/// — on either side of the iOS 17 divide — without a real system prompt. The
/// caller picks the tier: deciding which tier an ask resolves to is policy,
/// and policy lives in `PermissionService`, leaving implementations to do
/// nothing but report the OS's capabilities and dispatch to it.
protocol CalendarAuthorization {
  /// Whether this OS has the iOS 17+ write-only tier at all. The only version
  /// question `PermissionService` asks, and it asks it of the seam, so the
  /// pre-17 world is a test fixture rather than a runtime accident.
  var supportsWriteOnly: Bool { get }

  var status: OSAuthorization { get }

  /// Fires the OS prompt for `tier` and reports whether it was granted.
  func request(_ tier: CalendarPermissionType, completion: @escaping (_ granted: Bool) -> Void)
}

/// The production `CalendarAuthorization`, backed by EventKit. Every
/// `#available` check in the permission stack lives here.
struct EventKitAuthorization: CalendarAuthorization {
  let eventStore: EKEventStore

  var supportsWriteOnly: Bool {
    if #available(iOS 17.0, *) {
      return true
    }
    return false
  }

  var status: OSAuthorization {
    let status = EKEventStore.authorizationStatus(for: .event)

    if #available(iOS 17.0, *) {
      switch status {
      case .fullAccess:
        return .fullAccess
      case .writeOnly:
        return .writeOnly
      case .denied:
        return .denied
      case .restricted:
        return .restricted
      case .notDetermined:
        return .notDetermined
      @unknown default:
        return .denied
      }
    }

    // iOS 16 and below only has .authorized, which is full access.
    switch status {
    case .authorized:
      return .fullAccess
    case .denied:
      return .denied
    case .restricted:
      return .restricted
    case .notDetermined:
      return .notDetermined
    // Not `@unknown default`: the iOS 17+ `.writeOnly` case is unreachable
    // here but still counts against exhaustiveness, and a plain `default`
    // maps it to the same `.denied` without the warning.
    default:
      return .denied
    }
  }

  func request(_ tier: CalendarPermissionType, completion: @escaping (Bool) -> Void) {
    if #available(iOS 17.0, *) {
      switch tier {
      case .write:
        eventStore.requestWriteOnlyAccessToEvents { granted, _ in completion(granted) }
      case .full:
        eventStore.requestFullAccessToEvents { granted, _ in completion(granted) }
      }
    } else {
      // iOS 16 and below: only full access exists, and `PermissionService`
      // resolves every ask to `.full` there, so there is nothing to branch on.
      eventStore.requestAccess(to: .event) { granted, _ in completion(granted) }
    }
  }
}

/// The tier this app was granted, if any — see `record`.
///
/// On iOS 17+ `EKEventStore.authorizationStatus(for:)` can still report
/// `.notDetermined` for a short while after the request handler has already
/// confirmed a grant (#134), which made the very next call fail its permission
/// gate until the app was restarted. The recorded tier is the fallback the
/// `.notDetermined` branch consults.
///
/// Nothing ever clears it, and nothing needs to. It is only consulted while
/// the OS says `.notDetermined`, so any other status — a `.denied` from a
/// Settings revocation, say — wins outright. `.notDetermined` *is* itself a
/// status the OS can return to (Reset Location & Privacy, an MDM policy
/// change), and the record would mask that; what saves us is that iOS
/// terminates an app whose privacy settings change, so the record cannot
/// outlive the grant it describes.
///
/// Process-wide by design: `shared` is what production injects, because the
/// grant belongs to the app, not to a `PermissionService` — a
/// `FlutterEngineGroup` or add-to-app host builds one plugin (and so one
/// service) per engine, and an engine that did not itself ask must not fail
/// the gate during another engine's stale window. Tests inject a fresh
/// instance to stay isolated from each other.
final class GrantRecord {
  static let shared = GrantRecord()

  /// Read from the provider queue (the data endpoints) and from the main
  /// thread (the modal endpoints and the method-channel handlers), and written
  /// from whichever thread EventKit calls the request handler on, so every
  /// access goes through `lock`.
  private let lock = NSLock()
  private var granted: CalendarPermissionType?

  /// Remembers the tier the user just granted.
  func record(_ tier: CalendarPermissionType) {
    lock.lock()
    defer { lock.unlock() }
    // Only ever upgrade, so the record can never under-report what was granted.
    if granted == nil || tier == .full {
      granted = tier
    }
  }

  var current: CalendarPermissionType? {
    lock.lock()
    defer { lock.unlock() }
    return granted
  }
}

/// The permission policy: which tier an ask resolves to, which grants satisfy
/// which gates, and what Dart is told. It holds no version knowledge of its own
/// — the `CalendarAuthorization` seam answers that.
class PermissionService {
  private let authorization: CalendarAuthorization
  private let grantRecord: GrantRecord

  init(authorization: CalendarAuthorization, grantRecord: GrantRecord) {
    self.authorization = authorization
    self.grantRecord = grantRecord
  }

  /// The access this app effectively holds. Every decision — the gates, the
  /// status reported to Dart, and whether a request is worth firing — is made
  /// on this, so they can never disagree.
  ///
  /// The raw values are the wire format: they must match Dart's
  /// `CalendarPermissionStatus` enum, and this is their only definition.
  enum EffectiveAccess: String {
    case full = "granted"
    case writeOnly
    case denied
    case restricted
    /// Nothing granted and nothing recorded — we have not asked yet.
    case notDetermined

    init(granted type: CalendarPermissionType) {
      switch type {
      case .full:
        self = .full
      case .write:
        self = .writeOnly
      }
    }

    /// The tier lattice, defined once: full access covers everything, a
    /// write-only grant covers writes but never reads.
    func satisfies(_ required: CalendarPermissionType) -> Bool {
      switch required {
      case .full:
        return self == .full
      case .write:
        return self == .full || self == .writeOnly
      }
    }

    /// Can't be changed from inside the app — the user must use Settings.
    var isTerminal: Bool { self == .denied || self == .restricted }
  }

  /// Resolves the OS status, falling back to the recorded grant while — and
  /// only while — the OS still says `.notDetermined`. A real `.denied` (a
  /// Settings revocation, say) is honoured immediately and is never masked by
  /// an earlier grant.
  private func effectiveAccess() -> EffectiveAccess {
    switch authorization.status {
    case .fullAccess:
      return .full
    case .writeOnly:
      return .writeOnly
    case .denied:
      return .denied
    case .restricted:
      return .restricted
    case .notDetermined:
      return recordedAccess()
    }
  }

  private func recordedAccess() -> EffectiveAccess {
    grantRecord.current.map(EffectiveAccess.init(granted:)) ?? .notDetermined
  }

  /// Checks if calendar permissions are granted for the specified access level.
  ///
  /// - Parameter type: The type of access required (.write or .full)
  /// - Returns: true if the required permission level is granted
  func hasPermission(for type: CalendarPermissionType = .full) -> Bool {
    effectiveAccess().satisfies(type)
  }

  // Info.plist usage-description keys. The declaration checks and the error
  // messages must name the same keys, so both go through these constants and
  // the shared descriptionExample map.
  private static let legacyUsageKey = "NSCalendarsUsageDescription"
  private static let fullAccessUsageKey = "NSCalendarsFullAccessUsageDescription"
  private static let writeOnlyUsageKey = "NSCalendarsWriteOnlyAccessUsageDescription"

  private static let descriptionExamples: [String: String] = [
    legacyUsageKey: "Access your calendar to view and manage events.",
    fullAccessUsageKey: "Full access to view and edit your calendar events.",
    writeOnlyUsageKey: "Add events without reading existing events.",
  ]

  private func isDescriptionDeclared(_ key: String) -> Bool {
    let value = Bundle.main.object(forInfoDictionaryKey: key) as? String
    return !(value?.isEmpty ?? true)
  }

  private func missingDescriptionError(_ title: String, keys: [String]) -> PermissionError {
    var errorMessage = "\(title) not declared in Info.plist.\n\n"
    errorMessage += "Add the following to ios/Runner/Info.plist:"
    for key in keys {
      errorMessage += "\n<key>\(key)</key>"
      errorMessage += "\n<string>\(PermissionService.descriptionExamples[key] ?? "")</string>"
    }

    return PermissionError(code: PlatformExceptionCodes.permissionsNotDeclared, message: errorMessage)
  }

  private func missingWriteOnlyDescriptionError() -> PermissionError {
    missingDescriptionError(
      "Write-only calendar usage description",
      keys: [PermissionService.writeOnlyUsageKey])
  }

  private func missingFullAccessDescriptionError() -> PermissionError {
    missingDescriptionError(
      "Full-access calendar usage description",
      keys: [PermissionService.fullAccessUsageKey])
  }

  /// The no-keys-at-all error. Lists every key so following the advice once
  /// satisfies both the status guard and any later request on any OS version —
  /// naming only the legacy key would fix `hasPermissions` and then fail again
  /// on an iOS 17+ request, which demands the tier-specific keys.
  private func missingUsageDescriptionError() -> PermissionError {
    missingDescriptionError(
      "Calendar usage description",
      keys: [
        PermissionService.fullAccessUsageKey,
        PermissionService.writeOnlyUsageKey,
        PermissionService.legacyUsageKey,
      ])
  }

  /// Verifies the Info.plist declares the usage description the request needs.
  ///
  /// On iOS 17+ each request variant **requires** its own key: full access
  /// (`requestFullAccessToEvents`) needs `NSCalendarsFullAccessUsageDescription`
  /// and write-only needs `NSCalendarsWriteOnlyAccessUsageDescription` — the
  /// legacy `NSCalendarsUsageDescription` no longer satisfies either, and
  /// without the matching key the OS raises an exception, so we surface a
  /// clear error instead of crashing. iOS 16 and below uses only the legacy key.
  ///
  /// Takes the resolved tier rather than the caller's raw `writeOnly` flag, so
  /// the "which tier is this ask" question is answered exactly once, in
  /// `requestPermissions`.
  private func checkUsageDescriptionDeclared(for tier: CalendarPermissionType) -> PermissionError? {
    guard authorization.supportsWriteOnly else {
      return isDescriptionDeclared(PermissionService.legacyUsageKey)
        ? nil : missingDescriptionError(
          "Calendar usage description", keys: [PermissionService.legacyUsageKey])
    }

    switch tier {
    case .write:
      return isDescriptionDeclared(PermissionService.writeOnlyUsageKey)
        ? nil : missingWriteOnlyDescriptionError()
    case .full:
      return isDescriptionDeclared(PermissionService.fullAccessUsageKey)
        ? nil : missingFullAccessDescriptionError()
    }
  }

  func hasPermissions() -> Result<EffectiveAccess, PermissionError> {
    // A status check triggers no prompt, so any declared calendar usage
    // description — legacy, full-access, or write-only — satisfies the
    // configuration guard. An add-only app that declares only the write-only
    // key can still check status.
    guard isDescriptionDeclared(PermissionService.legacyUsageKey)
      || isDescriptionDeclared(PermissionService.fullAccessUsageKey)
      || isDescriptionDeclared(PermissionService.writeOnlyUsageKey) else {
      return .failure(missingUsageDescriptionError())
    }

    return .success(effectiveAccess())
  }

  /// Requests calendar access from the user.
  /// - Parameter writeOnly: when `true`, asks for add-only (write-only) access
  ///   where the OS supports it (iOS 17+). On iOS 16 and below write-only does
  ///   not exist, so the request falls back to full access regardless.
  func requestPermissions(
    writeOnly: Bool,
    completion: @escaping (Result<EffectiveAccess, PermissionError>) -> Void
  ) {
    // The tier this ask resolves to. Write-only exists only where the OS
    // supports it, so elsewhere every ask resolves to full access — the one
    // place the version difference is decided.
    let tier: CalendarPermissionType =
      writeOnly && authorization.supportsWriteOnly ? .write : .full

    let access = effectiveAccess()

    // Nothing to prompt for if we already hold a tier that satisfies the ask
    // (a full request while only write-only is held does *not*, so it falls
    // through), or if the answer can only be changed in Settings.
    if access.satisfies(tier) || access.isTerminal {
      completion(.success(access))
      return
    }

    // The usage-description key is only needed once we actually fire an OS
    // request, so the check sits after the early returns — an app that ships
    // without the (iOS 17+) tier key must still get its already-granted or
    // terminal status back rather than a configuration error.
    if let error = checkUsageDescriptionDeclared(for: tier) {
      completion(.failure(error))
      return
    }

    // Otherwise attempt the request. This prompts on a fresh notDetermined, and
    // also on a full request while only write-only is held — iOS re-presents the
    // dialog asking for full access and upgrades the app in-app if the user
    // agrees.
    authorization.request(tier) { granted in
      guard granted else {
        // Re-read rather than reporting denied outright, so a refused *full*
        // request while write-only is already held still reports the tier the
        // caller actually holds. Only when nothing resolves at all — the OS
        // still says notDetermined and nothing was recorded — do we answer
        // denied: the handler just told us the user said no, and reporting
        // "we never asked" would send callers down the wrong branch (#134's
        // stale window, pointing the other way).
        let access = self.effectiveAccess()
        completion(.success(access == .notDetermined ? .denied : access))
        return
      }
      // Record before replying: the caller's very next call may gate on this,
      // and the OS status can still read notDetermined at that point (#134).
      self.grantRecord.record(tier)
      completion(.success(EffectiveAccess(granted: tier)))
    }
  }
}

struct PermissionError: Error {
  let code: String
  let message: String
}
